### PR TITLE
Refactor Login

### PR DESCRIPTION
**Pull request by peacebonez**
Resolves https://github.com/uclaacm/TeachLAFrontend/issues/485

### CODE REVIEW

File: src/components/Login.js
Comment (lumisphere902) [inline src/components/Login.js:52]: Don't think this needs to be dependent on `dummy`

**Review by lumisphere902 — CHANGES_REQUESTED** [review]
LGTM w/ just one small nit

File: src/components/Login.js
Comment (peacebonez) [reply src/components/Login.js:52]: Copy that.  If not dependent on dummy, the dummy var is unused throughout the file.  Should I just set up lint-ignore to silence the warning that dummy is not used?

Comment (mattxwang) [reply src/components/Login.js:52]: Just to jump in here - I think the original intent of the code was to force a component re-render. If you can create a custom hook and/or some sort of listener that triggers a re-render (and then gets the window size, etc.), you can remove the dependency on the `dummy` state variable.

Comment (mattxwang) [reply src/components/Login.js:52]: Here's one common hook used for the same case: https://usehooks.com/useWindowSize/

Here's an interesting SO post on something similar; you may (?) have to use `useRef`: https://stackoverflow.com/questions/66588340/custom-hook-for-window-resize

File: src/components/Login.js
Comment (mattxwang) [inline src/components/Login.js:100]: Might be worth making `setIndex(Math.floor(Math.random() * 5))` a one-liner `const` function!

**Review by mattxwang — COMMENTED** [review]
Left some comments! I'm a bit far removed from the editor now, so feel free to merge without an approving review from me.

File: src/components/Login.js
Comment (peacebonez) [reply src/components/Login.js]: Hey Matt, have a look at the solution I have in place to dynamically render the SVG, it's looking quite nice when adjusting the window height.

Comment (peacebonez) [reply src/components/Login.js]: Ahh disregard, on larger screen heights the SVG appears buggy.  I'll have to take a closer look.

Screen Shot 2021-11-07 at 8 47 56 PM

Comment (peacebonez) [reply src/components/Login.js]: Ok I fixed this bug.  If the SVG was being rendered and then the innerHeight was increased, the SVG would look manipulated because it was not growing to fit the innerHeight.  Now the svg is growing with innerHeight.

File: src/components/Login.js
Comment (mattxwang) [inline src/components/Login.js:90]: I'm a bit confused as to why we need this state variable - don't all the state variables in `getBackgroundSVG` already rely on existing state, and thus automatically change on re-render? Right now, we're putting this JSX element into state, but it should be generated *from* state.

Comment (peacebonez) [reply src/components/Login.js:90]: When the svg is in state, the background svg is moving in an interesting wavy way that I thought was the intended effect but it's possible that it's a bug.  I'll remove the state and push up shortly

**Review by lumisphere902 — APPROVED** [review]
LGTM. Thanks for working on this; lots of good code quality upgrades. @mattxwang thoughts?

File: src/components/Login.js
Comment (mattxwang) [inline src/components/Login.js:92]: Is this the proper way to remove the event listener? I'm realizing that I may have made a mistake when I wrote the code initially. The two different instances of `() => setWindowHeight(window.innerHeight)` are anonymous functions with different underlying references, right? 

If that's the case, this code just needs to look like:

```suggestion
    const resizeHandler = setWindowHeight(window.innerHeight);
    window.addEventListener('resize', resizeHandler);
    return () => window.removeEventListener('resize', resizeHandler);
```

Comment (mattxwang) [reply src/components/Login.js:92]: Other than this, I'm good to merge. Not necessary to wait on me to approve as well, since Timothy's already given it the green light!

Comment (peacebonez) [reply src/components/Login.js]: This does the trick in getting the SVG to update in the intended way but there is a buggy side effect that I've experienced before - If the page is rendered and the window height is increased, you get a buggy looking SVG like the one in the screencap:
Screen Shot 2021-11-11 at 7 48 07 PM

With this code tweak:
```
const resizeHandler = () => setWindowHeight(window.innerHeight);
    window.addEventListener('resize', resizeHandler);
    return () => window.removeEventListener('resize', resizeHandler); 
```
It removes the bug, but the SVG doesn't have the cool swooping effect.

**Review by lumisphere902 — APPROVED** [review]
Reiterating my approval. Thanks for working on this!

**Comment by peacebonez** [comment]
Screen Shot 2021-11-13 at 1 21 35 PM
Hey guys, after rebasing I'm getting this linting error yelling at me about the function component.  Have you experienced this?  My local vs code is not indicating any linting errors.

**Review by mattxwang — DISMISSED** [review]

File: src/components/Login.js
Comment (mattxwang) [reply src/components/Login.js:44]: If you're curious, [here is the MDN doc](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions) that describes the differences from a normal function expression and an arrow function.

**Comment by mattxwang** [comment]
@peacebonez do you want to try running the ESLint autofixer locally with `npm run lint-js-fix`?

**Comment by lumisphere902** [comment]
It should automatically run it for you when you try to commit... have you ran `npm install`?

**Comment by mattxwang** [comment]
Oh also - @lumisphere902 why is the `package-lock.json` diff here so large? I would look at that before merging

**Comment by lumisphere902** [comment]
Ah, that could also be the source of errors. We are using Node v16.x

**Comment by peacebonez** [comment]
Hey guys sorry for the delay here (work + holidays), I'm going to run the suggested actions and get back to you.

**Comment by peacebonez** [comment]
All checks passed!  I'm now running node 16.2.0, generated a clean package-lock.json file, and followed the linting conventions when creating a new React Hook.

**Comment by mattxwang** [comment]
Just wanted to clarify - am I needed for review here? It's been a while since I've been heavily involved, but glad to lend a hand if needed.

**Comment by lumisphere902** [comment]
You're good; just going to merge this one after checks pass

**Comment by mattxwang** [comment]
Great work @peacebonez! This was a really hefty PR, and I think you've improved a lot of the bad code I wrote 3 years ago 😊

**Pull request merged**
